### PR TITLE
cephadm: Make list_networks ipv6 enabled

### PR DESCRIPTION
**Pull request by matthewoliver**
Currently the list_network command and methods in cephadm only run and
parse ipv4 output from `ip route`.

This patch extends the list_network command and internal methods to be
ipv6 enabled. It now also checks `ip -6 route` and `ip -6 addr` to
gather gather all networks from both protocol families.

Signed-off-by: Matthew Oliver <moliver@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -2418,7 +2418,8 @@ def command_bootstrap():
         # make sure IP is configured locally, and then figure out the
         # CIDR network
         for net, ips in list_networks().items():
-            if base_ip in ips:
+            if ipaddress.ip_address(unicode(base_ip)) in \
+                    [ipaddress.ip_address(unicode(ip)) for ip in ips]:
```
Comment (matthewoliver) [inline src/cephadm/cephadm:2422]: Ipv6 address string could come in shortened in different way. So we use ipaddress to normalise it. The unicode is for py2 support.

File: src/cephadm/cephadm [hunk 2]
```diff
@@ -3138,6 +3144,36 @@ def _parse_ip_route(out):
...
+            r[net] = []
+
+    for line in ips.splitlines():
+        m = ip_p.findall(line)
+        if not m:
+            continue
+        ip = m[0][0]
+        # find the network it belongs to
+        net = [n for n in r.keys()
+               if ipaddress.ip_address(unicode(ip)) in ipaddress.ip_network(unicode(n))]
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:3171]:
```
In [5]: out = !ip -6 route ls

In [6]: out
Out[6]: 
['2003:0:0:0::/64 dev enxd89ef3f34260 proto ra metric 100 pref medium',
 '2003:0:0:0::/56 via fe80::e228:6dff:fef3:84dc dev enxd89ef3f34260 proto ra metric 100 pref medium',
 '2620:0:0:0:10:163:0:2 dev tun0 proto kernel metric 50 pref medium',
 '2620:0:0:0:10:163:0:2 dev tun0 proto kernel metric 256 pref medium',
 '2620:0:0:0:10:163:1:179 dev tun0 proto kernel metric 50 pref medium',
 '2620:0:0:8000::/50 via 2620:0:0:0:10:163:0:2 dev tun0 proto static metric 50 pref medium',
 'fe80::/64 dev enxd89ef3f34260 proto kernel metric 100 pref medium',
 'fe80::/64 dev enxd89ef3f34260 proto kernel metric 256 pref medium',
 'default dev tun0 proto static metric 50 pref medium',
 'default via fe80::e228:6dff:fef3:84dc dev enxd89ef3f34260 proto ra metric 100 pref medium']

In [8]: ips = !ip -6 addr ls

In [9]: ips
Out[9]: 
['1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 state UNKNOWN qlen 1000',
 '    inet6 ::1/128 scope host ',
 '       valid_lft forever preferred_lft forever',
 '12: enxd89ef3f34260: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 state UP qlen 1000',
 '    inet6 2003:0:0:0:fdda:781e:968f:996b/64 scope global temporary dynamic ',
 '       valid_lft 7183sec preferred_lft 1298sec',
 '    inet6 2003:0:0:0:8d47:225:a2c9:e943/64 scope global dynamic mngtmpaddr noprefixroute ',
 '       valid_lft 7183sec preferred_lft 1298sec',
 '    inet6 fe80::e6a8:490a:9144:aa1/64 scope link noprefixroute ',
 '       valid_lft forever preferred_lft forever',
 '15: tun0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 state UNKNOWN qlen 100',
 '    inet6 2620:0:0:0:10:163:1:179 peer 2620:0:0:0:10:163:0:2/64 scope global noprefixroute ',
 '       valid_lft forever preferred_lft forever',
 '    inet6 fe80::758b:ac49:f114:40d0/64 scope link stable-privacy ',
 '       valid_lft forever preferred_lft forever']


In [16]: _parse_ipv6_route('\n'.join(out), '\n'.join(ips))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-16-fc22267d9e72> in <module>()
----> 1 _parse_ipv6_route('\n'.join(out), '\n'.join(ips))

<ipython-input-1-5ef68d3a6c08> in _parse_ipv6_route(routes, ips)
     17         ip = m[0][0]
     18         # find the network it belongs to
---> 19         net = [n for n in r.keys()
     20                if ipaddress.ip_address(unicode(ip)) in ipaddress.ip_network(unicode(n))]
     21         if net:

<ipython-input-1-5ef68d3a6c08> in <listcomp>(.0)
     18         # find the network it belongs to
     19         net = [n for n in r.keys()
---> 20                if ipaddress.ip_address(unicode(ip)) in ipaddress.ip_network(unicode(n))]
     21         if net:
     22             r[net[0]].append(ip)

/usr/lib/python3.6/ipaddress.py in ip_network(address, strict)
     82 
     83     raise ValueError('%r does not appear to be an IPv4 or IPv6 network' %
---> 84                      address)
     85 
     86 

ValueError: 'default' does not appear to be an IPv4 or IPv6 network
```

Comment (matthewoliver) [reply src/cephadm/cephadm:3171]: oh yeah, not filtering out default, thanks Sebastian!

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-07-06_04:02:43-rados-wip-kefu-testing-2020-07-06-1016-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-07-06_11:39:50-rados-wip-kefu-testing-2020-07-06-1016-distro-basic-smithi/

**Pull request merged**